### ISSUE TITLE
Add tests for ensuring loss goes down

### ISSUE THREAD

**Issue by Rocamonde** [HumanCompatibleAI/imitation#562]
## Bug description
Currently we are mostly not testing for the loss going down in different algorithms. Recently we noticed that we had forgotten to add a `loss.backward()` call on #481 on `preference_comparisons.BasicRewardTrainer()`, which was not picked up by tests. This would have rendered the algorithm basically useless. 

To prevent this from happening in the future, we should develop a sensible test suite to ensure that optimizers are bringing the loss down, and in general that algorithms are progressing as they should.

A bug like this going undetected can compromise the reliability of our benchmarks and the software, so IMO it's important to make sure this is well tested. 

AFAIK we currently kind of do this through the `reward_improvement` module for BC, DAgger, and density, but not more generally and thoroughly, certainly not for preference comparisons.

Tests should also be layered, as it's pretty hard to identify in a complex algorithm that has multiple optimizers why it suddenly stopped improving - it's much easier to fix e.g. that the loss of a specific component is not decreasing at all.

It could also be a good idea to put a lot of the tooling within the package to allow other users to run these sanity benchmarks in their custom modifications or implementations.

**Comment by AdamGleave** [comment]
Fixed in #601
